src/registrar/models/domain.py

```diff
@@ -267,25 +373,234 @@ def administrative_contact(self) -> PublicContact:
+            )
+            # TODO-error handling better here?
+
+    def _set_singleton_contact(self, contact: PublicContact, expectedType: str):  # noqa
+        """Sets the contacts by adding them to the registry as new contacts,
+        updates the contact if it is already in epp,
+        deletes any additional contacts of the matching type for this domain
+        does not create the PublicContact object, this should be made beforehand
+        (call save() on a public contact to trigger the contact setters
+        which inturn call this function)
+        Raises ValueError if expected type doesn't match the contact type"""
+        if expectedType != contact.contact_type:
+            raise ValueError(
+                "Cannot set a contact with a different contact type,"
+                " expected type was %s" % expectedType
+            )
+
+        isRegistrant = contact.contact_type == contact.ContactTypeChoices.REGISTRANT
+        isEmptySecurity = (
+            contact.contact_type == contact.ContactTypeChoices.SECURITY
+            and contact.email == ""
+        )
+
+        # get publicContact objects that have the matching
+        # domain and type but a different id
+        # like in highlander we there can only be one
+        hasOtherContact = (
+            PublicContact.objects.exclude(registry_id=contact.registry_id)
+            .filter(domain=self, contact_type=contact.contact_type)
+            .exists()
+        )
+
+        # if no record exists with this contact type
+        # make contact in registry, duplicate and errors handled there
+        errorCode = self._make_contact_in_registry(contact)
+
+        # contact is already added to the domain, but something may have changed on it
+        alreadyExistsInRegistry = errorCode == ErrorCode.OBJECT_EXISTS
+        # if an error occured besides duplication, stop
+        if (
+            not alreadyExistsInRegistry
+            and errorCode != ErrorCode.COMMAND_COMPLETED_SUCCESSFULLY
+        ):
+            # TODO- ticket #433 look here for error handling
+            raise Exception("Unable to add contact to registry")
+
+        # contact doesn't exist on the domain yet
+        logger.info("_set_singleton_contact()-> contact has been added to the registry")
+
+        # if has conflicting contacts in our db remove them
+        if hasOtherContact:
+            logger.info(
+                "_set_singleton_contact()-> updating domain, removing old contact"
+            )
+            if isEmptySecurity:
+                existing_contact = PublicContact.objects.filter(
+                    domain=self, contact_type=contact.contact_type
+                ).get()
+            else:
+                existing_contact = (
+                    PublicContact.objects.exclude(registry_id=contact.registry_id)
+                    .filter(domain=self, contact_type=contact.contact_type)
+                    .get()
+                )
+            if isRegistrant:
```

Comment:
By storing filtered contacts in a variable as described above (`filteredContacts = PublicContact.objects.filter(domain=self, contact_type=contact.contact_type`), this also allows you to refactor these as so (ignore the weird github formatting... may need to correct it):
```suggestion
            if isEmptySecurity:
                   filteredContacts.get()
            else:
                existing_contact = (
                    filteredContacts.exclude(registry_id=contact.registry_id).get()
                )
            if isRegistrant:
```
You could take this further if you want and store the .exclude operation in a variable, but this should prevent .filter being called twice, as we should only need to do that once here